server/client.go

```diff
@@ -769,11 +773,26 @@ func (c *client) writeLoop() {
 		if close = c.flags.isSet(closeConnection); !close {
 			owtf := c.out.fsp > 0 && c.out.pb < maxBufSize && c.out.fsp < maxFlushPending
 			if waitOk && (c.out.pb == 0 || owtf) {
+				fspBeforeWait := c.out.fsp
+				c.mu.Unlock()
+
+				var timeout bool
+				// Set our maximum wait to 3secs.
+				t.Reset(3 * time.Second)
+
 				// Wait on pending data.
-				c.out.sgw = true
-				c.out.sg.Wait()
-				c.out.sgw = false
+				select {
+				case <-ch:
+				case <-t.C:
+					timeout = true
+				}
+
+				c.mu.Lock()
 				close = c.flags.isSet(closeConnection)
+				if !close && timeout && fspBeforeWait > 0 && time.Since(lastReport) > 10*time.Minute {
```

Comment:
To start I think we log all of them IMO.

Reply:
Afraid that there will be either no print at all or we will spam the log :-)

Reply:
I think we will get the edge condition then it will clear. Let's take risk, more data is better to start with IMO.